src/server/pfs/s3/s3.go

```diff
@@ -109,20 +110,46 @@ func Server(port uint16, driver Driver, clientFactory ClientFactory) (*http.Serv
 	s3Server.Bucket = c
 	s3Server.Object = c
 	s3Server.Multipart = c
-	router := s3Server.Router()
+	return s3Server.Router()
+}
+
+// Server runs an HTTP server with an S3-like API for PFS. This allows you to
+// use s3 clients to access PFS contents.
 
+// Modifications to `routerMap` allow dynamic changes for the mapping between s3 endpoints and API configurations
+func Server(port uint16, defaultRouter *mux.Router, routerMap map[string]*mux.Router, routersLock *sync.Mutex) *http.Server {
+	logger := logrus.WithFields(logrus.Fields{
+		"source": "s3gateway",
+	})
 	server := &http.Server{
 		Addr:         fmt.Sprintf(":%d", port),
 		ReadTimeout:  requestTimeout,
 		WriteTimeout: requestTimeout,
 		Handler: http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
 			// Log that a request was made
 			logger.Infof("http request: %s %s", r.Method, r.RequestURI)
-			router.ServeHTTP(w, r)
+			if strings.HasPrefix(r.Host, "s3-") {
+				routersLock.Lock()
```

Comment:
a RWMutex could make sense here, but I doubt it makes much difference

Reply:
i agree with both statements :)

i'll switch to it since it seems logical